gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+        for (RrsetWrapper wrapper : zone.dnsRecords().get(zone.zone().getName())) {
+          ResourceRecordSet wrappedRrset = wrapper.rrset();
+          if (rrset.getName().equals(wrappedRrset.getName())
+              && rrset.getType().equals(wrappedRrset.getType())) {
+            // such a record exist and we must have a deletion
+            if (deletions == null || !deletions.contains(wrappedRrset)) {
+              return Error.ALREADY_EXISTS.response(String.format(
+                  "The 'entity.change.additions[%s]' resource named '%s (%s)' already exists.",
+                  index, rrset.getName(), rrset.getType()));
+            }
+          }
+        }
+        if (rrset.getType().equals("SOA") && findByNameAndType(deletions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(deletions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    if (deletions != null) {
+      int index = 0;
+      for (ResourceRecordSet rrset : deletions) {
+        if (rrset.getType().equals("SOA") && findByNameAndType(additions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(additions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Helper for searching rrsets in a collection.
+   */
+  private static ResourceRecordSet findByNameAndType(Iterable<ResourceRecordSet> records,
+      String name, String type) {
+    if (records != null) {
+      for (ResourceRecordSet rrset : records) {
+        if ((name == null || name.equals(rrset.getName()))
+            && (type == null || type.equals(rrset.getType()))) {
+          return rrset;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * We only provide the most basic validation for A and AAAA records.
+   */
+  static boolean checkRrData(String data, String type) {
```

Comment:
I'm not exactly how strict cloud dns is in input format.  Do you think parts of this method could be replaced with
[isValidInetAddress](
http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/net/InetAddresses.html#isInetAddress(java.lang.String)) (implementation [here](https://github.com/google/guava/blob/master/guava/src/com/google/common/net/InetAddresses.java))?